### PR TITLE
From dataset136

### PR DESCRIPTION
**Pull request by cmacdonald**
Not finished, but could be reviewed.

### CODE REVIEW

File: pyterrier/batchretrieve.py [hunk 1]
```diff
@@ -62,11 +63,70 @@ def __init__(self, verbose=0, **kwargs):
...
+    import json
+    
+    if isinstance(dataset, str):
+        dataset = get_dataset(dataset)
+    if version != "latest":
+        raise ValueError("index versioning not yet supported")
+    indexref = dataset.get_index(variant)
+
+    classname = clz.__name__
+    # now look for, e.g., BatchRetrieve.args.json file, which will define the args for BatchRetrieve, e.g. stemming
```
Comment (seanmacavaney) [inline pyterrier/batchretrieve.py:84]: Will naming this BatchRetrieve.args.json make it annoying if/when we change the name of the class later?

File: pyterrier/batchretrieve.py [hunk 2]
```diff
@@ -62,11 +63,70 @@ def __init__(self, verbose=0, **kwargs):
...
+    indexdir = indexref #os.path.dirname(indexref.toString())
+    argsfile = os.path.join(indexdir, classname + ".args.json")
+    if os.path.exists(argsfile):
+        with autoopen(argsfile, "rt") as f:
+            args = json.load(f)
+            # anything specified in kwargs of this methods overrides the .args.json file
+            for k,v in kwargs.items():
```
Comment (seanmacavaney) [inline pyterrier/batchretrieve.py:91]: `args.update(kwargs)`?

File: pyterrier/batchretrieve.py [hunk 3]
```diff
@@ -62,11 +63,70 @@ def __init__(self, verbose=0, **kwargs):
...
+                args[k] = v
+            kwargs = args
+    return clz(indexref, **kwargs)   
+                
 class BatchRetrieve(BatchRetrieveBase):
     """
     Use this class for retrieval by Terrier
     """
 
+    #from_dataset = staticmethod(partial(_from_dataset, clz=BatchRetrieve))
```
Comment (seanmacavaney) [inline pyterrier/batchretrieve.py:101]: cleanup (also, the `partial` import at the top)

File: pyterrier/datasets.py [hunk 4]
```diff
@@ -201,17 +203,51 @@ def _get_one_file(self, component, variant=None):
...
-        for (local, URL) in file_list:
+
+        direxists = os.path.exists(localDir)
+        
+        location = self.locations[component]
+        if is_lambda(location) or isinstance(location, types.FunctionType):
+            # functions are expensive to call, normally another HTTP is needed.
+            # just assume we have everthing we need if we have the local directory already
+            if direxists:
+                return localDir
```
Comment (seanmacavaney) [inline pyterrier/datasets.py:222]: Should most of this function be wrapped in a try/finally to clean up the dir if anything goes wrong?

Comment (cmacdonald) [reply pyterrier/datasets.py:222]: I was wondering about touching an empty file showing that download completion has occurred, and checking if that exists.

File: pyterrier/batchretrieve.py (hunk 1 above)
Comment (cmacdonald) [reply pyterrier/batchretrieve.py:84]: I think we can just make an additional json file and run that for a while.

File: pyterrier/batchretrieve.py [hunk 5]
```diff
@@ -98,8 +96,7 @@ class BatchRetrieve(BatchRetrieveBase):
     Use this class for retrieval by Terrier
     """
 
-    #from_dataset = staticmethod(partial(_from_dataset, clz=BatchRetrieve))
-    @staticmethod
+        @staticmethod
```
Comment (seanmacavaney) [inline pyterrier/batchretrieve.py:99]: indentation looks off here? (is this tabs vs spaces?)

File: pyterrier/datasets.py (hunk 4 above)
Comment (seanmacavaney) [reply pyterrier/datasets.py:222]: that sounds less ugly, yeah

File: pyterrier/batchretrieve.py (hunk 5 above)
Comment (cmacdonald) [reply pyterrier/batchretrieve.py]: yes, I spotted. and ran unit tests before pushing again. this should be fixed now.

**Review by seanmacavaney — APPROVED** [review]

**Review by seanmacavaney — APPROVED** [review]

**Pull request merged**